index.js

```diff
@@ -142,57 +94,77 @@ module.exports = function (opts) {
-        timeout_handshake = timeout_inactivity = opts.timeout
+      // needed for tests
+      if (!opts.connections)
+        opts.connections = {
+          incoming: {
+            net: [{ scope: "public", "transform": "shs" }]
+          },
+          outgoing: {
+            net: [{ transform: "shs" }]
+          }
+        }
+
+      var peers = api.peers = {}
 
-      var shsCap = (opts.caps && opts.caps.shs) || opts.appKey || appKey
-      var shs = Shs({
-        keys: opts.keys && toSodiumKeys(opts.keys),
-        seed: opts.seed,
-        appKey: toBuffer(shsCap),
+      var transports = []
 
-        //****************************************
-        timeout: timeout_handshake,
+      var transforms = [
+        //function () { return shs }
+      ]
 
-        authenticate: function (pub, cb) {
-          var id = '@'+u.toId(pub)
-          api.auth(id, function (err, auth) {
-            if(err) cb(err)
-            else    cb(null, auth || create.permissions.anonymous)
+      var server, ms, ms_client
+
+      function setupMultiserver () {
+        if(server) return server
+        if(transforms.length < 1)
+          throw new Error('secret-stack needs at least 1 transform protocol')
+
+        var server_suites = []
+        var client_suites = []
+
+        for (var incTransportType in opts.connections.incoming) {
+          opts.connections.incoming[incTransportType].forEach(function (conf) {
+            transforms.forEach(function (transform) {
+              transports.forEach(function (transport) {
+                if (transport.name == incTransportType && transform.name == conf.transform)
+                  server_suites.push([
+                    transport.create(conf),
+                    transform.create()
+                  ])
+              })
+            })
           })
         }
-      })
-
-      //figure out the local key. take this from the shs plugin,
-      //because we may have only passed in a seed.
-      var id = '@'+u.toId(shs.publicKey)
-
-      //use configured port, or a random user port.
-      var port = opts.port || 1024+(~~(Math.random()*(65536-1024)))
-      var host = opts.host || nonPrivate.v4 || nonPrivate.private.v4 || '127.0.0.1'
 
-      var peers = api.peers = {}
-
-      var server_protocols = [
-        [Net({port: port, host: host}), shs],
-        [Onion({server: false}), shs]
-      ]
-      var client_protocols = server_protocols
+        for (var outTransportType in opts.connections.outgoing) {
+          opts.connections.outgoing[outTransportType].forEach(function (conf) {
+            transforms.forEach(function (transform) {
+              transports.forEach(function (transport) {
+                if (transport.name == outTransportType && transform.name == conf.transform)
+                  client_suites.push([
+                    transport.create(conf),
+                    transform.create()
+                  ])
+              })
+            })
+          })
+        }
```

Comment:
Logic looks great. Do you know which versions of JavaScript we're supporting? If we can use `Object.values()` we can replace the top two loops (here and elsewhere) with:

```js
const confs = Object.values(opts.connections.outgoing);
Object.values(confs).forEach(function (conf) {
```

Not sure whether it's in scope for this PR, but we could also use a Cartesian product and replace *all* of the nested loops with something like:

```js
var confs = Object.values(opts.connections.outgoing)
cartesian(confs, transforms, transports).forEach(function (conf, transform, transport) {
  if (transport.name == outTransportType && transform.name == conf.transform)
    client_suites.push([
      transport.create(conf),
      transform.create()
    ])
})
```

I think lodash and friends can do Cartesian products, but I [found](https://stackoverflow.com/a/43053803) a two-liner (in ES6):

```js
const f = (a, b) => [].concat(...a.map(d => b.map(e => [].concat(d, e))));
const cartesian = (a, b, ...c) => (b ? cartesian(f(a, b), ...c) : a);
```

Reply:
I like the less nested and shorter code but that two-liner also isn't easy to understand but a clear name and a comment should help with that, I think...

AFAICT `Object.values` works in node 8. I'm not sure about 6 and how long we want to support it.

Reply:
@christianbundy backend stuff tries to target es5

Reply:
from the SO link:

> My example uses the spread operator and the rest parameters

According to the mozilla MDN, spread is supported since node 5 and rest param since node 6. So I think this would be fine to since we only test node v6 and upwards anyway?

@christianbundy can you link me to the _elsewhere_ you mentioned? Depending on the number of cases we might want to do this in a separate PR?